File: JavaLibrary/src/test/java/com/flat/simplex/parser/logic/block/BlockReturnTest.java
package com.flat.simplex.parser.logic.block;

import com.flat.simplex.parser.logic.Context;
import com.flat.simplex.parser.logic.error.Error;
import com.flat.simplex.support.TokenChain;
import org.junit.jupiter.api.Test;

import static com.flat.simplex.support.ContextSupport.assertErrors;
import static com.flat.simplex.support.TokenChain.mChain;
import static com.flat.simplex.support.TokenChain.readChain;
import static org.junit.jupiter.api.Assertions.*;

class BlockReturnTest {

    @Test
    public void blockBreak() {
        TokenChain chain = readChain("return;");

        Context context = new Context(chain.get());
        BlockReturn block = new BlockReturn(context, chain.get(), null);
        block.read();

        assertNull(block.getTokenContent());
        assertNull(block.getTokenContentEnd());

        assertErrors(context);
    }

    @Test
    public void blockBreakContent() {
        TokenChain chain = readChain("return hello;");

        Context context = new Context(chain.get());
        BlockReturn block = new BlockReturn(context, chain.get(), null);
        block.read();

        TokenChain.assertChain("hello;", block.getTokenContent(), block.getTokenContentEnd(), "Invalid body");

        assertErrors(context);
    }

    @Test
    public void blockBreakUnexpectedToken_Fail() {
        TokenChain chain = readChain("return;;");

        Context context = new Context(chain.get());
        BlockReturn block = new BlockReturn(context, chain.get(), null);
        block.read();

        assertNull(block.getTokenContent());
        assertNull(block.getTokenContentEnd());

        assertErrors(context, Error.unexpectedToken);
    }

    @Test
    public void blockBreakUnexpectedEndOfTokens_Fail() {
        TokenChain chain = readChain("return");

        Context context = new Context(chain.get());
        BlockReturn block = new BlockReturn(context, chain.get(), null);
        block.read();

        assertNull(block.getTokenContent());
        assertNull(block.getTokenContentEnd());

        assertErrors(context, Error.unexpectedEndOfTokens);
    }
}

File: SimplexApp/simplex/parser/logic/block/BlockReturn.test.js
const TokenChain = require("simplex/support/TokenChain.js");
const Context = require("simplex/parser/logic/Context.js");
const Error = require("simplex/parser/logic/error/Error.js");
const BlockReturn = require("simplex/parser/logic/block/BlockReturn.js")
const ContextSupport = require("simplex/support/ContextSupport.js");
const AssertError = require("simplex/support/AssertError.js");

const readChain = TokenChain.readChain;
const assertErrors = ContextSupport.assertErrors;
const assertNull = AssertError.assertNull;

test('blockBreak', () => {
    let chain = readChain('return;');
    let context = new Context(chain.get());
    let block = new BlockReturn(context, chain.get(), null);
    block.read();
    assertNull(block.getTokenContent());
    assertNull(block.getTokenContentEnd());
    assertErrors(context);
});
test('blockBreakContent', () => {
    let chain = readChain('return hello;');
    let context = new Context(chain.get());
    let block = new BlockReturn(context, chain.get(), null);
    block.read();
    TokenChain.assertChain('hello;', block.getTokenContent(), block.getTokenContentEnd(), 'Invalid body');
    assertErrors(context);
});
test('blockBreakUnexpectedToken_Fail', () => {
    let chain = readChain('return;;');
    let context = new Context(chain.get());
    let block = new BlockReturn(context, chain.get(), null);
    block.read();
    assertNull(block.getTokenContent());
    assertNull(block.getTokenContentEnd());
    assertErrors(context, Error.unexpectedToken);
});
test('blockBreakUnexpectedEndOfTokens_Fail', () => {
    let chain = readChain('return');
    let context = new Context(chain.get());
    let block = new BlockReturn(context, chain.get(), null);
    block.read();
    assertNull(block.getTokenContent());
    assertNull(block.getTokenContentEnd());
    assertErrors(context, Error.unexpectedEndOfTokens);
});